django/db/models/fields/tuple_lookups.py

```diff
@@ -34,52 +38,68 @@ def check_rhs_length_equals_lhs_length(self):
                 f"must have {len_lhs} elements"
             )
 
-    def as_sql(self, compiler, connection):
-        # e.g.: (a, b, c) == (x, y, z) as SQL:
-        # WHERE (a, b, c) = (x, y, z)
-        vals = [
+    def get_prep_lhs(self):
+        if isinstance(self.lhs, (tuple, list)):
+            return Tuple(*self.lhs)
+        else:
+            return super().get_prep_lhs()
+
+    def process_lhs(self, compiler, connection, lhs=None):
+        sql, params = super(BuiltinLookup, self).process_lhs(compiler, connection, lhs)
+        if not sql.startswith("("):
+            sql = f"({sql})"
+        return sql, params
+
+    def process_rhs(self, compiler, connection):
+        values = [
             Value(val, output_field=col.output_field)
             for col, val in zip(self.lhs, self.rhs)
         ]
-        lookup_class = self.__class__.__bases__[-1]
-        lookup = lookup_class(Tuple(self.lhs), Tuple(*vals))
-        return lookup.as_sql(compiler, connection)
+
+        return Tuple(*values).as_sql(compiler, connection)
 
 
 class TupleExact(TupleLookupMixin, Exact):
     def as_oracle(self, compiler, connection):
         # e.g.: (a, b, c) == (x, y, z) as SQL:
         # WHERE a = x AND b = y AND c = z
-        cols = self.lhs.get_cols()
-        lookups = [Exact(col, val) for col, val in zip(cols, self.rhs)]
+        lookups = [Exact(col, val) for col, val in zip(self.lhs, self.rhs)]
         root = WhereNode(lookups, connector=AND)
 
         return root.as_sql(compiler, connection)
 
 
-class TupleIsNull(IsNull):
-    def as_sql(self, compiler, connection):
-        # e.g.: (a, b, c) is None as SQL:
-        # WHERE a IS NULL AND b IS NULL AND c IS NULL
-        vals = self.rhs
-        if isinstance(vals, bool):
-            vals = [vals] * len(self.lhs)
+class TupleIsNull(TupleLookupMixin, IsNull):
+    def get_prep_lookup(self):
+        rhs = self.rhs
 
-        cols = self.lhs.get_cols()
-        lookups = [IsNull(col, val) for col, val in zip(cols, vals)]
-        root = WhereNode(lookups, connector=AND)
+        if isinstance(rhs, (tuple, list)) and len(rhs) == 1:
+            rhs = rhs[0]
+        if isinstance(rhs, bool):
+            return rhs
 
+        raise ValueError(
+            "The QuerySet value for an isnull lookup must be True or False."
+        )
```

Comment:
It's actually already tested in `test_isnull_non_boolean_value`

Reply:
I can delete the code and that test still passes

Reply:
Hmm, yes you can delete it and it will pass. That's because `IsNull.as_sql` is called from `TupleIsNull.as_sql`. If instead of deleting it you modify the error message, you'll see the test fails.